core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionLoader.java

```diff
@@ -164,7 +164,68 @@ protected void loadAllExtensions(@NonNull ExtensionManager extensionManager) {
             Map<String, GeyserExtensionContainer> loadedExtensions = new LinkedHashMap<>();
 
             Pattern[] extensionFilters = this.extensionFilters();
-            List<Path> extensionPaths = Files.walk(extensionsDirectory).toList();
+
+            Path updateDirectory = extensionsDirectory.resolve("update");
+            List<Path> extensionPaths;
+            if (Files.isDirectory(updateDirectory)) {
+                // Get the current extensions and store them in a map
+                Map<String, Path> extensionFiles = new HashMap<>();
+                extensionPaths = Files.list(extensionsDirectory).toList();
+                extensionPaths.forEach(path -> {
+                    if (Files.isDirectory(path)) {
+                        return;
+                    }
+
+                    // Only look at files that meet the extension filter
+                    for (Pattern filter : extensionFilters) {
+                        if (!filter.matcher(path.getFileName().toString()).matches()) {
+                            return;
+                        }
+                    }
+
+                    try {
+                        // Try load the description, so we know it's a valid extension
+                        GeyserExtensionDescription description = this.extensionDescription(path);
+
+                        // Store the file name against ID for later use
+                        extensionFiles.put(description.id(), path);
+                    } catch (Throwable e) {
+                        // no-op
+                    }
+                });
+
+                // Perform the updates
+                List<Path> extensionUpdatePaths = Files.list(updateDirectory).toList();
+                extensionUpdatePaths.forEach(path -> {
+                    if (Files.isDirectory(path)) {
+                        return;
+                    }
+
+                    // Only look at files that meet the extension filter
+                    for (Pattern filter : extensionFilters) {
+                        if (!filter.matcher(path.getFileName().toString()).matches()) {
+                            return;
+                        }
+                    }
+
+                    try {
+                        // Try load the description, so we know it's a valid extension
+                        GeyserExtensionDescription description = this.extensionDescription(path);
+
+                        // Remove the old extension with the same ID if it exists
+                        if (extensionFiles.containsKey(description.id())) {
+                            Files.delete(extensionFiles.get(description.id()));
+                        }
```

Comment:
nitpick, but this could probably be one `.get(description.id())` call, which is then null-checked

Further, I wonder what would happen if there are multiple update jars for one extension (e.g. 2 different versions) in the update folder